### ISSUE TITLE
UIDropDownMenu taint can break edit mode and actionbars

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#299]
UIDropDownMenu frames can taint edit mode, propagating this to action bars and then breaking them in similar ways seen in #298 such as in-combat paging and use of keybindings to activate abilities.

# Test Case

## Preconditions

1. Place a few abilities on Action Bar 1 and ensure they are keybound.
2. Import the following layout into Edit Mode. The only change to this layout from the "Modern" preset is increasing the icon padding on Action Bar 1 from 2 to 3.
3. Ensure the current Edit Mode layout is set to "Modern".

```
0 39 0 0 1 7 7 UIParent 0.0 45.0 -1 ##$$%/&('&)#+# 0 1 1 7 7 UIParent 0.0 45.0 -1 ##$$%/&('%(#,$ 0 2 1 7 7 UIParent 0.0 45.0 -1 ##$$%/&('%(#,$ 0 3 1 5 5 UIParent -5.0 -77.0 -1 #$$$%/&('%(#,$ 0 4 1 5 5 UIParent -5.0 -77.0 -1 #$$$%/&('%(#,$ 0 5 1 1 4 UIParent 0.0 0.0 -1 ##$$%/&('%(#,$ 0 6 1 1 4 UIParent 0.0 -50.0 -1 ##$$%/&('%(#,$ 0 7 1 1 4 UIParent 0.0 -100.0 -1 ##$$%/&('%(#,$ 0 10 1 7 7 UIParent 0.0 45.0 -1 ##$$&('% 0 11 1 7 7 UIParent 0.0 45.0 -1 ##$$&('%,# 0 12 1 7 7 UIParent 0.0 45.0 -1 ##$$&('% 1 -1 1 4 4 UIParent 0.0 0.0 -1 ##$#%# 2 -1 1 2 2 UIParent 0.0 0.0 -1 ##$#%( 3 0 1 8 7 UIParent -300.0 250.0 -1 $#3# 3 1 1 6 7 UIParent 300.0 250.0 -1 %#3# 3 2 1 6 7 UIParent 520.0 265.0 -1 %#&#3# 3 3 1 0 2 CompactRaidFrameManager 0.0 -7.0 -1 '#(#)#-#.#/#1$3# 3 4 1 0 2 CompactRaidFrameManager 0.0 -5.0 -1 ,#-#.#/#0#1#2( 3 5 1 5 5 UIParent 0.0 0.0 -1 &#*$3# 3 6 1 5 5 UIParent 0.0 0.0 -1 -#.#/#4& 3 7 1 4 4 UIParent 0.0 0.0 -1 3# 4 -1 1 7 7 UIParent 0.0 45.0 -1 # 5 -1 1 7 7 UIParent 0.0 45.0 -1 # 6 0 1 2 2 UIParent -255.0 -10.0 -1 ##$#%#&.(()( 6 1 1 2 2 UIParent -270.0 -155.0 -1 ##$#%#'+(()( 7 -1 1 7 7 UIParent 0.0 45.0 -1 # 8 -1 0 6 6 UIParent 35.0 50.0 -1 #'$A%$&7 9 -1 1 7 7 UIParent 0.0 45.0 -1 # 10 -1 1 0 0 UIParent 16.0 -116.0 -1 # 11 -1 1 8 8 UIParent -9.0 85.0 -1 # 12 -1 1 2 2 UIParent -110.0 -275.0 -1 #K$# 13 -1 1 8 8 MicroButtonAndBagsBar 0.0 0.0 -1 ##$#%#&- 14 -1 1 2 2 MicroButtonAndBagsBar 0.0 10.0 -1 ##$#%( 15 0 1 7 7 StatusTrackingBarManager 0.0 0.0 -1 # 15 1 1 7 7 StatusTrackingBarManager 0.0 17.0 -1 # 16 -1 1 5 5 UIParent 0.0 0.0 -1 #( 17 -1 1 1 1 UIParent 0.0 -100.0 -1 ## 18 -1 1 5 5 UIParent 0.0 0.0 -1 #- 19 -1 1 7 7 UIParent 0.0 0.0 -1 ##
```

## Steps

This bug has a precondition in that Edit Mode must be entered through a means other than the right-click unit popup menus on the player frame or the chat tab.

1. Execute the following chat command to create a dropdown menu in the middle of the screen.
```
/run local D = CreateFrame("Frame", nil, UIParent, "UIDropDownMenuTemplate") D:SetPoint("CENTER") UIDropDownMenu_SetInitializeFunction(D, function() UIDropDownMenu_AddButton({ text = "foo" }) end)
```
2. Open the dropdown menu in the middle of the screen and close the menu again.
4. Run `/editmode`.
5. Change the current active layout to the imported layout.
6. Cycle your main action bar page up and down one time.
7. Use any ability on your main action bar via a keybinding.

Upon using the ability via a keybinding you'll be met with an action blocked popup. Additionally, many interactions with action bars in-combat will no longer work - such as trying to change the current page, and dragging and dropping buttons. You'll also get script errors upon auto-attacking enemies.

# Explanation

This taint is caused through the following sequence of events.

Firstly, opening the addon-created dropdown and running its insecure initialize function calls [UIDropDownMenu_AddButton]. The [DropDownList1.numButtons], and [DropDownList1.maxWidth] fields will be written during this call with tainted values.

When the dropdown menu is closed, these fields on `DropDownList1` remain will remain tainted. They are only reset to secure values if [UIDropDownMenu_InitializeHelper] is called (via [UIDropDownMenu_Initialize]).

When opening edit mode the following sequence of calls occurs:

1. [EditModeManagerFrame:EnterEditMode]
2. [EditModeManagerFrame:UpdateDropdownOptions] - This updates options for the LayoutDropdown child frame.
3. [DropDownControlMixin:SetOptions]
8. [DropDownControlMixin:Initialize]
9. [UIDropDownMenu_AddButton]

At no point in this sequence of calls is it confirmed that the layout dropdown is actually open; the [UIDropDownMenu_AddButton] function is called unconditionally and will append to the *currently or last opened* dropdown list, which will lead to [a read of our previously tainted fields](https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/SharedXML/UIDropDownMenu.lua#L398) as no other dropdown has been properly initialized via [UIDropDownMenu_Initialize] prior to entering edit mode to reset them. This is shown in the following taint log excerpt, which then spreads taint to the rest of Edit Mode.

```
10/10 11:32:27.803  Execution tainted by *** ForceTaint_Strong *** while reading numButtons - Interface/SharedXML/UIDropDownMenu.lua:398 UIDropDownMenu_AddButton()
10/10 11:32:27.803      Interface/SharedXML/SharedUIPanelTemplates.lua:2411 Frame:Initialize()
10/10 11:32:27.803      Interface/SharedXML/SharedUIPanelTemplates.lua:2492 Frame:SetOptions()
10/10 11:32:27.803      Interface/FrameXML/EditModeManager.lua:976 EditModeManagerFrame:UpdateDropdownOptions()
10/10 11:32:27.803      Interface/FrameXML/EditModeManager.lua:169 EditModeManagerFrame:EnterEditMode()
```

# Possible Solutions

A simple but untested solution to resolve this would probably be for either [DropDownControlMixin:Initialize] or [DropDownControlMixin:SetOptions] to validate that the dropdown is actually presently open (or in the process of being initialized) before trying to add entries to the list via [UIDropDownMenu_AddButton]. 

Alternatively, explicitly intializing the dropdown via a call to UIDropDownMenu_Initialize would clear all the tainted fields being read.

[EditModeManagerFrame:EnterEditMode]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/FrameXML/EditModeManager.lua#L166
[EditModeManagerFrame:UpdateDropdownOptions]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/FrameXML/EditModeManager.lua#L921
[DropDownControlMixin:SetOptions]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/SharedXML/SharedUIPanelTemplates.lua#L2490
[DropDownControlMixin:Initialize]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/SharedXML/SharedUIPanelTemplates.lua#L2372
[UIDropDownMenu_Initialize]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/SharedXML/UIDropDownMenu.lua#L70-L73
[UIDropDownMenu_InitializeHelper]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/SharedXML/UIDropDownMenu.lua#L34-L49
[UIDropDownMenu_AddButton]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/SharedXML/UIDropDownMenu.lua#L386
[DropDownList1.numButtons]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/SharedXML/UIDropDownMenu.lua#L409
[DropDownList1.maxWidth]: https://github.com/Gethe/wow-ui-source/blob/15a65c6cfd23400d23c012c6ee4f1375e4b16abd/Interface/SharedXML/UIDropDownMenu.lua#L696-L697

**Comment by timothy003** [comment]
Can't reproduce on live (10.1.0.50000), but I'm still seeing action bars randomly getting broken, sometimes in the middle of a key. Can you please swap the calls to `SetButtonState()` and `TryUseActionButton()` in [ActionButtonDown()](https://github.com/Gethe/wow-ui-source/blob/e8693cc8bc1f4ca2391216badfcf63290328aefd/Interface/FrameXML/ActionButton.lua#L124-L128) and [ActionButtonUp()](https://github.com/Gethe/wow-ui-source/blob/e8693cc8bc1f4ca2391216badfcf63290328aefd/Interface/FrameXML/ActionButton.lua#L140-L141) so that action bars don't break so easily?

**Comment by Meorawr** [comment]
It's still an issue - though it doesn't raise a blocked action error immediately upon entering into Edit Mode now, and requires (at least) changing the Edit Mode layout to one that slightly modifies aspects of action bars.

I've updated the instructions accordingly.

**Comment by timothy003** [comment]
Still can't reproduce this. How does edit mode taint action bars? I'm seeing `UpdateAction()` execution being tainted by `wipe()` (likely from [`UpdateShownButtons()`](https://github.com/Gethe/wow-ui-source/blob/e8693cc8bc1f4ca2391216badfcf63290328aefd/Interface/FrameXML/ActionBar.lua#L189)) immediately upon changing talents or specialization. I don't use edit mode.

**Comment by Meorawr** [comment]
Make sure you've not got any addons enabled when testing this.

**Comment by timothy003** [comment]
Whoops! Somehow, WeakAuras was preventing the taint from happening. With it disabled, I get an action blocked popup immediately upon entering edit mode.

```
6/28 04:12:00.420  Global variable SELECTED_DOCK_FRAME tainted by *** ForceTaint_Strong *** - Interface/FrameXML/FloatingChatFrame.lua:1582 FCF_SelectDockFrame()
6/28 04:12:00.420      Interface/FrameXML/EditModeSystemTemplates.lua:1733 ChatFrame1:OnEditModeEnter()
6/28 04:12:00.420      Interface/FrameXML/EditModeManager.lua:173
6/28 04:12:00.420      secureexecuterange()
6/28 04:12:00.420      Interface/FrameXML/EditModeManager.lua:177 EditModeManagerFrame:ShowSystemSelections()
6/28 04:12:00.420      Interface/FrameXML/EditModeManager.lua:184 EditModeManagerFrame:EnterEditMode()
6/28 04:12:00.420      Interface/FrameXML/EditModeManager.lua:210
6/28 04:12:00.420      EditModeManagerFrame:Show()
6/28 04:12:00.420      Interface/FrameXML/UIParent.lua:2943 Frame:SetUIPanel()
6/28 04:12:00.420      Interface/FrameXML/UIParent.lua:2747 Frame:ShowUIPanel()
6/28 04:12:00.420      Interface/FrameXML/UIParent.lua:2651
6/28 04:12:00.420      Frame:SetAttribute()
6/28 04:12:00.420      Interface/FrameXML/UIParent.lua:3253 ShowUIPanel()
6/28 04:12:00.420      Interface/FrameXML/ChatFrame.lua:2712 ?()
6/28 04:12:00.420      Interface/FrameXML/ChatFrame.lua:5236 ChatEdit_ParseText()
6/28 04:12:00.420      Interface/FrameXML/ChatFrame.lua:4894 ChatEdit_SendText()
6/28 04:12:00.420      Interface/FrameXML/ChatFrame.lua:4930 ChatEdit_OnEnterPressed()
6/28 04:12:00.420      ChatFrame1EditBox:hatFrame.xml:141_OnEnterPressed()
6/28 04:12:00.424  Interface/SharedXML/UIDropDownMenu.lua:413 UIDropDownMenu_AddButton()
6/28 04:12:00.424  An action was blocked because of taint from *** ForceTaint_Strong *** - TargetUnit()
6/28 04:12:00.424      Interface/FrameXML/EditModeManager.lua:1722 Frame:RefreshTargetAndFocus()
6/28 04:12:00.424      Interface/FrameXML/EditModeManager.lua:1636 Frame:OnEditModeEnter()
6/28 04:12:00.424      Interface/FrameXML/EditModeManager.lua:185 EditModeManagerFrame:EnterEditMode()
6/28 04:12:00.424      Interface/FrameXML/EditModeManager.lua:210
6/28 04:12:00.424      EditModeManagerFrame:Show()
6/28 04:12:00.424      Interface/FrameXML/UIParent.lua:2943 Frame:SetUIPanel()
6/28 04:12:00.424      Interface/FrameXML/UIParent.lua:2747 Frame:ShowUIPanel()
6/28 04:12:00.424      Interface/FrameXML/UIParent.lua:2651
6/28 04:12:00.424      Frame:SetAttribute()
6/28 04:12:00.424      Interface/FrameXML/UIParent.lua:3253 ShowUIPanel()
6/28 04:12:00.424      Interface/FrameXML/ChatFrame.lua:2712 ?()
6/28 04:12:00.424      Interface/FrameXML/ChatFrame.lua:5236 ChatEdit_ParseText()
6/28 04:12:00.424      Interface/FrameXML/ChatFrame.lua:4894 ChatEdit_SendText()
6/28 04:12:00.424      Interface/FrameXML/ChatFrame.lua:4930 ChatEdit_OnEnterPressed()
6/28 04:12:00.424      ChatFrame1EditBox:hatFrame.xml:141_OnEnterPressed()
6/28 04:12:00.424  Interface/SharedXML/UIDropDownMenu.lua:413 UIDropDownMenu_AddButton()
6/28 04:12:00.424  An action was blocked because of taint from *** ForceTaint_Strong *** - FocusUnit()
6/28 04:12:00.424      Interface/FrameXML/EditModeManager.lua:1726 Frame:RefreshTargetAndFocus()
6/28 04:12:00.424      Interface/FrameXML/EditModeManager.lua:1636 Frame:OnEditModeEnter()
6/28 04:12:00.424      Interface/FrameXML/EditModeManager.lua:185 EditModeManagerFrame:EnterEditMode()
6/28 04:12:00.424      Interface/FrameXML/EditModeManager.lua:210
6/28 04:12:00.424      EditModeManagerFrame:Show()
6/28 04:12:00.424      Interface/FrameXML/UIParent.lua:2943 Frame:SetUIPanel()
6/28 04:12:00.424      Interface/FrameXML/UIParent.lua:2747 Frame:ShowUIPanel()
6/28 04:12:00.424      Interface/FrameXML/UIParent.lua:2651
6/28 04:12:00.424      Frame:SetAttribute()
6/28 04:12:00.424      Interface/FrameXML/UIParent.lua:3253 ShowUIPanel()
6/28 04:12:00.424      Interface/FrameXML/ChatFrame.lua:2712 ?()
6/28 04:12:00.424      Interface/FrameXML/ChatFrame.lua:5236 ChatEdit_ParseText()
6/28 04:12:00.424      Interface/FrameXML/ChatFrame.lua:4894 ChatEdit_SendText()
6/28 04:12:00.424      Interface/FrameXML/ChatFrame.lua:4930 ChatEdit_OnEnterPressed()
6/28 04:12:00.424      ChatFrame1EditBox:hatFrame.xml:141_OnEnterPressed()
```

Still not sure how this can happen without edit mode, though.

**Comment by Nevcairiel** [comment]
Edit Mode gets invokved no matter if you open it specifically or not, since it manages the position of frames, even if you leave them in the default locations.

**Comment by Meorawr** [comment]
This appears to be resolved in 11.0, which now validates the dropdown before attempting to populate it _and_ routes through UIDropDownMenu_Initialize rather than blindly adding buttons to the last-opened menu.
